src/dotnet/GeneratorAgent/src/Azure.Tools.GeneratorAgent/GitHubTypeSpecSdkGenerationService.cs

```diff
@@ -0,0 +1,175 @@
+        private readonly string SdkOutputDirectory;
+
+        public GitHubTypeSpecSdkGenerationService(
+            AppSettings appSettings,
+            ILogger<GitHubTypeSpecSdkGenerationService> logger,
+            ProcessExecutor processExecutor,
+            string commitId,
+            string typespecSpecDirectory,
+            string sdkOutputDirectory)
+        {
+            ArgumentNullException.ThrowIfNull(appSettings);
+            ArgumentNullException.ThrowIfNull(logger);
+            ArgumentNullException.ThrowIfNull(processExecutor);
+            ArgumentException.ThrowIfNullOrWhiteSpace(commitId);
+            ArgumentException.ThrowIfNullOrWhiteSpace(typespecSpecDirectory);
+            ArgumentException.ThrowIfNullOrWhiteSpace(sdkOutputDirectory);
+
+            AppSettings = appSettings;
+            Logger = logger;
+            ProcessExecutor = processExecutor;
+            CommitId = commitId;
+            TypespecSpecDirectory = typespecSpecDirectory;
+            SdkOutputDirectory = sdkOutputDirectory;
+        }
+
+        public async Task<bool> CompileTypeSpecAsync(CancellationToken cancellationToken = default)
+        {
+            try
+            {
+                Logger.LogInformation("Starting GitHub-based TypeSpec compilation for commit: {CommitId}", CommitId);
+                Logger.LogInformation("SDK output directory: {SdkOutputDirectory}", SdkOutputDirectory);
+                Logger.LogInformation("TypeSpec spec directory: {TypeSpecSpecDirectory}", TypespecSpecDirectory);
+
+                // Step 1: Extract azure-sdk-for-net root path
+                string azureSdkPath = ExtractAzureSdkPath();
+
+                // Step 2: Run PowerShell generation script
+                if (!await RunPowerShellGenerationScript(azureSdkPath, cancellationToken))
+                {
+                    return false;
+                }
+
+                // Step 3: Navigate to src directory and run dotnet build
+                if (!await RunDotNetBuildGenerateCode(cancellationToken))
+                {
+                    return false;
+                }
+                return true;
+            }
+            catch (Exception ex)
+            {
+                Logger.LogError(ex, "Unexpected error during GitHub-based TypeSpec compilation");
+                return false;
+            }
+        }
+
+        protected virtual string ExtractAzureSdkPath()
+        {
+            // Extract azure-sdk-for-net path from SDK output directory
+            string azureSdkPath = SdkOutputDirectory;
+            while (!string.IsNullOrEmpty(azureSdkPath) && !Path.GetFileName(azureSdkPath).Equals(AppSettings.AzureSdkDirectoryName, StringComparison.OrdinalIgnoreCase))
+            {
+                string? parentPath = Path.GetDirectoryName(azureSdkPath);
+                if (parentPath == null)
+                {
+                    break;
+                }
+                azureSdkPath = parentPath;
```

Comment:
nit: I think you could simplify this to:
```suggestion
                azureSdkPath = Path.GetDirectoryName(azureSdkPath);
```

The while loop will terminate naturally when `azureSdkPath` is null or empty and the check below will still handle it as before.